### ISSUE TITLE
Bug: [no-floating-promises] Can't create array of promises and pass to Promise.race

### ISSUE THREAD

**Issue by dcecile** [typescript-eslint/typescript-eslint#8180]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=IYZwngdgxgBAZgV2gFwJYHsLwBQgFwwDKApsgDwAKATugLaojFkBu6qAJgHycCUMA3gCgYMYAHdgqZDGp0GxAHRVgUYtgDaCrSAXEIyKqmIhsPALo8A3IIC%2BgwaEixEKDFgDmuArPqMWbLnULAWFRCSkZGl9FZVVcK1tBIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
async function f(s: Set<Promise<void>>) {
  await Promise.race([...s.entries()]);
}

async function g(s: Promise<void>[]) {
  await Promise.race(s);
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-floating-promises": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Both `f` and `g` should be OK

### Actual Result

`f` is has an error at `Promise.race`

> An array of Promises may be unintentional. Consider handling the promises' fulfillment or rejection with Promise.all or similar, or explicitly marking the expression as ignored with the `void` operator. 2:3 - 2:40

### Additional Info

[Docs](https://typescript-eslint.io/rules/no-floating-promises/) list 4 special methods:

```
Promise.all(),
Promise.allSettled(),
Promise.any()
Promise.race()
```

(But I didn't [find any](https://github.com/search?q=repo%3Atypescript-eslint%2Ftypescript-eslint%20allSettled&type=code) code references in the repo to e.g. `allSettled`, and didn't find `race` in [no-floating-promises.ts](https://github.com/typescript-eslint/typescript-eslint/blob/v6.17.0/packages/eslint-plugin/src/rules/no-floating-promises.ts))

**Comment by bradzacher** [comment]
This is working as expected - you're using the wrong method on the `Set`.

`.entries()` returns `Iterator<[T, T]>` - i.e each element returned by the iterator is a 2 element tuple where both elements are the value.

To show this more clearly:
```ts
const x = new Set(['a', 'b', 'c']);
for (const e of x.entries()) {
  console.log(e);
}
// prints:
// ['a', 'a']
// ['b', 'b']
// ['c', 'c']
```

So for `Set<Promise<void>>`, `.entries` returns `Iterator<[Promise<void>, Promise<void>]>`.
So `[...s.entries()]` is of type `Array<[Promise<void>, Promise<void>]>`.

`Promise.race` - like the other batch promise waiter methods - will just pass through non-promise values. In particular `.race` will just return the first non-promise value (as it's the first "promise" to resolve). So that means that `Promise.race([[Promise.resolve(), Promise.resolve()]])` just returns the tuple - wrapped in a promise.

So to show you what this means:
```ts
declare const set: Set<Promise<void>>;
const promisesWRONG = [...set.entries()];
//    ^? Array<[Promise<void>, Promise<void>]>
const raceResultWRONG = Promise.race(promisesWRONG);
//    ^? Promise<[Promise<void>, Promise<void>]>
const awaitedWrong = await raceResultWRONG;
//    ^? [Promise<void>, Promise<void>]
```

Instead use `.values()` on your set:

```ts
declare const set: Set<Promise<void>>;
const promisesRIGHT = [...set.values()];
//    ^? Array<Promise<void>>
const raceResultRIGHT = Promise.race(promisesRIGHT);
//    ^? Promise<void>
const awaitedRIGHT = await raceResultRIGHT;
//    ^? void
```

As you can see by the types - this is the correct method to use.

**Comment by dcecile** [comment]
Oh! Thanks, that absolutely makes sense, glad that the lint caught this
